include/nlohmann/detail/input/input_adapters.hpp

```diff
@@ -132,30 +133,46 @@ class iterator_input_adapter
     using char_type = typename std::iterator_traits<IteratorType>::value_type;
 
     iterator_input_adapter(IteratorType first, IteratorType last)
-        : current(std::move(first)), end(std::move(last))
+        : current(std::move(first)), end(std::move(last)), current_has_been_consumed(false)
     {}
 
     typename std::char_traits<char_type>::int_type get_character()
     {
-        if (JSON_HEDLEY_LIKELY(current != end))
+
+        if (JSON_HEDLEY_LIKELY(current_has_been_consumed))
         {
-            auto result = std::char_traits<char_type>::to_int_type(*current);
             std::advance(current, 1);
-            return result;
         }
 
-        return std::char_traits<char_type>::eof();
+        if (JSON_HEDLEY_LIKELY(current != end))
+        {
+            current_has_been_consumed = true;
+            return std::char_traits<char_type>::to_int_type(*current);
+        }
+        else
+        {
+            current_has_been_consumed = false;
+            return std::char_traits<char_type>::eof();
+        }
+
     }
 
   private:
     IteratorType current;
     IteratorType end;
+    bool current_has_been_consumed = false;
 
     template<typename BaseInputAdapter, size_t T>
     friend struct wide_string_input_helper;
 
-    bool empty() const
+    bool empty()
```

Comment:
Should this remain `const` qualified? (Meaning `current` and `current_has_been_consumed` would have to be made `mutable`.)

Reply:
Do it should?
Empty actually need to do work (advance the underlying `iterator`) if input has been already consumed

keeping it `const` would be only to legacy reasons

Reply:
Well, that's why we have the `mutable` keyword in the language. It's more to meet expectations. You're not going to expect `empty()` to not work on a `const` object based on how the same function operates in other classes.

Reply:
Roger